pairing/bls12381/point.go

```diff
@@ -0,0 +1,403 @@
+}
+
+func (k *pointG2) Add(a, b kyber.Point) kyber.Point {
+	aa := a.(*pointG2)
+	bb := b.(*pointG2)
+	bls12381.NewG2().Add(k.p, aa.p, bb.p)
+	return k
+}
+
+func (k *pointG2) Sub(a, b kyber.Point) kyber.Point {
+	aa := a.(*pointG2)
+	bb := b.(*pointG2)
+	bls12381.NewG2().Sub(k.p, aa.p, bb.p)
+	return k
+}
+
+func (k *pointG2) Neg(a kyber.Point) kyber.Point {
+	aa := a.(*pointG2)
+	bls12381.NewG2().Neg(k.p, aa.p)
+	return k
+}
+
+func (k *pointG2) Mul(s kyber.Scalar, q kyber.Point) kyber.Point {
+	if q == nil {
+		q = nullpointG2().Base()
+	}
+	bls12381.NewG2().MulScalarBig(
+		k.p,
+		q.(*pointG2).p,
+		&s.(*mod.Int).V,
+	)
+	return k
+}
+
+func (k *pointG2) MarshalBinary() ([]byte, error) {
+	return bls12381.NewG2().ToCompressed(k.p), nil
+}
+
+func (k *pointG2) UnmarshalBinary(buff []byte) error {
+	var err error
+	k.p, err = bls12381.NewG2().FromCompressed(buff)
+	return err
+}
+
+func (k *pointG2) MarshalTo(w io.Writer) (int, error) {
+	buf, err := k.MarshalBinary()
+	if err != nil {
+		return 0, err
+	}
+	return w.Write(buf)
+}
+
+func (k *pointG2) UnmarshalFrom(r io.Reader) (int, error) {
+	buf := make([]byte, k.MarshalSize())
+	n, err := io.ReadFull(r, buf)
+	if err != nil {
+		return n, err
+	}
+	return n, k.UnmarshalBinary(buf)
+}
+
+func (k *pointG2) MarshalSize() int {
+	return 96
+}
+
+func (k *pointG2) String() string {
+	b, _ := k.MarshalBinary()
+	return "bls12-381.G1: " + hex.EncodeToString(b)
+}
+
+func (k *pointG2) Hash(m []byte) kyber.Point {
+	pg2, _ := bls12381.NewG2().HashToCurve(m, Domain)
+	k.p = pg2
+	return k
+}
+
+func sha256Hash(in []byte) []byte {
+	h := sha256.New()
+	h.Write(in)
+	return h.Sum(nil)
+}
+
+func (k *pointG2) IsInCorrectGroup() bool {
+	return bls12381.NewG2().InCorrectSubgroup(k.p)
+}
+
+type pointGT struct {
+	f *bls12381.E
+}
+
+func newEmptyGT() *pointGT {
+	return newPointGT(bls12381.NewGT().New())
+}
+func newPointGT(f *bls12381.E) *pointGT {
+	return &pointGT{
+		f: f,
+	}
+}
+
+func (k *pointGT) Equal(kk kyber.Point) bool {
+	return k.f.Equal(kk.(*pointGT).f)
+}
+
+const gtLength = 576
+
+func (k *pointGT) Null() kyber.Point {
+	var zero [gtLength]byte
+	k.f, _ = bls12381.NewGT().FromBytes(zero[:])
+	return k
+}
+
+func (k *pointGT) Base() kyber.Point {
+	panic("not yet available")
+}
+
+func (k *pointGT) Pick(rand cipher.Stream) kyber.Point {
+	panic("TODO: bls12-381.GT.Pick()")
+}
+
+func (k *pointGT) Set(q kyber.Point) kyber.Point {
+	k.f.Set(q.(*pointGT).f)
+	return k
+}
+
+func (k *pointGT) Clone() kyber.Point {
+	kk := newEmptyGT()
+	kk.Set(k)
+	return kk
+}
+
+func (k *pointGT) Add(a, b kyber.Point) kyber.Point {
+	aa := a.(*pointGT)
+	bb := b.(*pointGT)
+	bls12381.NewGT().Mul(k.f, aa.f, bb.f)
+	return k
+}
+
+func (k *pointGT) Sub(a, b kyber.Point) kyber.Point {
+	aa := a.(*pointGT)
+	bb := b.(*pointGT)
+	bls12381.NewGT().Sub(k.f, aa.f, bb.f)
+	return k
+}
+
+func (k *pointGT) Neg(q kyber.Point) kyber.Point {
+	panic("bls12-381: GT is not a full kyber.Point implementation")
```

Comment:
This and the next function are available in `bls12-381`, so I would implement them.